### ISSUE TITLE
Optic Category

### ISSUE THREAD

**Issue by tysonzero** [well-typed/optics#279]
Currently `Optic` uses 4 type variables for `s t a b` and thus is not a category. However conceptually `Optic` seems like a pretty ideal `Category`. So I figured I'd at least bring up the possibility.

This would essentially involve changing the kind of `Optic` from:

```
OpticKind -> IxList -> Type -> Type -> Type -> Type -> Type
```

to:

```
OpticKind -> IxList -> (Type, Type) -> (Type, Type) -> Type
```

I think the majority of the type signatures either don't directly manipulate `s` and `t`, or they manipulate them simultaneously / in-sync (e.g. `%`). So I actually think most type signatures would be simplified by this change.

As mentioned this would allow us to give `Optic` a `Category` instance, and thus `>>>` / category's `.` could be used for composition. It would also allow any existing functionality based around `Category` to seamlessly be used on `Optic`.

Currently `%` allows for some type changing outside of the `s t a b` changes, which could not be reflected in `>>>`. I have not use `optics` enough to know how problematic this would be, and I would propose keeping `%` around either way. I am also unsure if it's possible to use more lens-style polymorphism to handle this, where the lenses have all the polymorphism, and the call sites are where the lenses are made concrete.

I know this is a very large change, so it might not be realistic or desirable, but I figured it couldn't hurt to mention the idea.

**Comment by phadej** [comment]
`>>>` would be very problematic, you don't compose `_1 >>> traversed`. Optic-kind changing thing is very common

**Comment by tysonzero** [comment]
Ah yes I see. Would someone be able to give me a quick summary behind the reasoning for making the lenses themselves monomorphic, but the combinators that use them polymorphic, as opposed to `lens`'s approach which is the opposite. It seems like this would avoid a lot of the `Is` / explicit supercasting stuff.

**Comment by phadej** [comment]
I think @adamgundry could comment better on that choice, I don't have a quick answer. For me it follows from the fact, that if you try to define an **opaque** `Optic` type, then if you try to also have monomorphic combinators with polymorphic optics, you'll have hard time getting it all to work.

I invite you to try yourself to make it work:

```haskell
-- this isn't really opaque, but ok it's not transparent either
type Optic k s a = forall token. k token => OpticLike token s a 
newtype Optic token s a = Optic ...
```

the token would probably be the same `A_Lens`, `A_Traversal` enumeration.

---

As a particular example, note that `optics` don't form full lattice, we don't have all combinations:

```haskell
>>> sets map % to not
...
...A_Setter cannot be composed with A_Getter
```

See https://hackage.haskell.org/package/optics-0.2/docs/Optics.html#g:4

**Comment by tysonzero** [comment]
> I invite you to try yourself to make it work

Now i'm very curious, so i'll think I'll have to give it a go.

> As a particular example, note that optics don't form full lattice, we don't have all combinations

That's not particularly problematic. Composing typeclasses can lead to a type that is uninhabited, for example some types implement `(Monad, Comonad)`, and some types implement `Contravariant`, but if you need a single type to both then you are out of luck.

**Comment by tysonzero** [comment]
[Here](https://pastebin.com/3pbycyHj) is what I came up with.

It's simplified and doesn't deal with type changing updates / prisms / indexing. However it should hopefully give a good idea for how this kind of thing would work.

If you wanted to consolidate all the types under a single `Optic` type you would be perfectly able to. I used separate simple concrete types for simplicity.

Your `sets map >>> to not` example would have the type `(FromGetter cat, FromSetter cat, Category cat) => cat [Bool] Bool`, which no type implements, so it would fail to typecheck regardless of which combinator you tried to use it with.

**Comment by phadej** [comment]
That is nice, you even have comprehensible type-errors:

```haskell
*Optics> Nothing ^. _Just

<interactive>:1:12: error:
    • No instance for (FromTraversal Getter)
```

and reasonable type inference

```haskell
Optics> :t _1 >>> _Just
_1 >>> traversed
  :: (Category cat, FromTraversal cat, Traversable t) =>
     cat (t c, b) c
```

Note: it says `FromTraversal`, because your `_Just` is traversal, not prism. It would say

```haskell
  :: (Category cat, FromLens cat, FromPrism cat, Traversable t) =>
```

otherwise.

---

Compare that with optics:

```haskell
*Optics> view _Just Nothing

<interactive>:1:1: error:
    • A_Prism cannot be used as A_Getter

*Optics> :t _1 % traversed
_1 % _Just
  :: Field1 s t (Maybe a) (Maybe b) =>
     Optic An_AffineTraversal '[] s t a b
```

`optics` say directly that it's `An_AffineTraversal`. By using concrete kinds, we can compute *lower* bound. Users don't need to refer to the graph in a documentation to find out what's combination of `Lens` and `Prism`.

---

Your design looks promising, but it's up to you to develop it.

**Comment by phadej** [comment]
Another example which works with `optics` is following:

```haskell
foo :: Lens' A Int -> A -> A
foo l x | view l x == 0 = x               -- using as getter
        | otherwise     = over l succ x   -- using as setter
```

you'd need to add explicit `fromLens` calls.

So there are trade-offs to make between whether actual optics or combinators
are monomorphic in a optic kind. I cannot say that one is clearly better than
another.

I still vaguely remember that @adamgundry had very good point about this,
but somehow I cannot recall it.

**Comment by tysonzero** [comment]
Yeah I honestly don't know where I fall on this ultimately. I do have to say I quite like the elegance of `Category` and not having to define an explicit `meet` operation, but monomorphic lenses are definitely easier to pass around as you don't need `RankNTypes` or explicit `from*` calls.

**Comment by tysonzero** [comment]
> Your design looks promising, but it's up to you to develop it.

Thanks, i'm glad it seemed reasonable to you. At some point I will try and do so, particularly depending on what ends up happening with `.foo.bar` desugaring.

It also leads to a somewhat interesting and different approach where you build in the opposite direction.

You start with the most powerful lens, so presumably something like `Equality` or `Iso`, and then you weaken from there. This is somewhat atypical, although it honestly doesn't sound so bad. It could even allow you to build an impure hierarchy on top of the pure hierarchy, as opposed to the opposite which most would probably not be willing to do.

**Comment by arybczak** [comment]
Hmm, this looks interesting. It seems equally powerful, there are just different tradeoffs.

> It seems like this would avoid a lot of the Is / explicit supercasting stuff to follow in lens's footsteps here.

Doesn't that just change their names and move them to a different place? Instead of `Is A_Lens A_Traversal` you need to have `FromLens Traversal` instance (which would be pretty much the same, i.e. `id` modulo rewrapping newtypes).

**Comment by tysonzero** [comment]
@arybczak 

> It seems equally powerful, there are just different tradeoffs.

Yeah that sounds right to me.

It seems like composition is nicer with the polymorphic lens approach, but passing around and storing lenses is nicer with the monomorphic lens approach. There are probably some other tradeoffs too.

> Doesn't that just change their names and move them to a different place? Instead of Is A_Lens A_Traversal you need to have FromLens Traversal instance

Yeah you are right. Either way you essentially have `Is` under some name or another, either on the producer of the lens side or on the consumer of the lens side.

However it does seem like you get to avoid needing `Join` for composition. Since the end use of the lens monomorphizes everything in the chain immediately to the needed type.
